Feature/reports server side

kulish-alina opened the pull request:
#328

Commit:
ReportService was created

Report models were implemented

kulish-alina committed:
add two methods in ReportsController;

add VacanciesReportParameters

kulish-alina committed:
add UsersReportDTO.cs;

add VacanciesReportDTO;

Commit:
ReportService under construction

kulish-alina committed:
property Creator was added in Candidate;

kulish-alina committed:
property LastModifiedDate was added to Vacancy

kulish-alina committed:
new state was added in Vacancy;

property StartDate is not required;
new field LastModifiedDate for vacancy state was added

kulish-alina committed:
VacancyState was added to project

kulish-alina committed:
StatesInfo was added to Vacancy

kulish-alina committed:
ReportService under construction

kulish-alina committed:
method for getting data for vacancies report was added to ReportService

kulish-alina committed:
method for DailyVacanciesReport was implemented in ReportService

kulish-alina committed:
some small corrections in ReportService;

ReportsController was implemented;

kulish-alina committed:
small corrections

kulish-alina committed:
property LastModifiedDate was removed from vacancy;

mykytayefrueiv commented on src/BaseOfTalents/DAL/DTO/ReportDTO/DailyVacanciesReportDTO.cs:
```diff
@@ -0,0 +1,17 @@
+    {
+        public int UserId { get; set; }
+        public string DisplayName { get; set; }
+        public int PendinVacanciesCount { get; set; }
```
Pendin**g**VacanciesCount

mykytayefrueiv commented on src/BaseOfTalents/DAL/DTO/ReportDTO/UsersReportDTO.cs:
```diff
@@ -0,0 +1,21 @@
+        public int UserId { get; set; }
+        public string DisplayName { get; set; }
+
+        public Dictionary<int, int> StagesData { get; set; } //key stage Id (added = id-0), vslue - data from db
```
v**a**lue

mykytayefrueiv replied:
Not really understand about what "data" you writting

kulish-alina committed:
comments were added to ReportService;

selection and filtering vacancies' states from data base were moved  to private methods;

msvaillant commented on src/BaseOfTalents/DAL/Services/ReportService.cs:
```diff
@@ -0,0 +1,189 @@
+
+            var result = new List<LocationsUsersReportDTO>();
+
+            foreach (var usersInCity in usersGroup)
```
Maybe multiple looping can be replaced with querying, consider some changes

msvaillant commented on src/BaseOfTalents/DAL/Services/ReportService.cs:
```diff
@@ -0,0 +1,189 @@
+        {
+            var result = new Dictionary<int, int>();
+            var data = stages.Where(s => s.Vacancy.ResponsibleId == userId).GroupBy(x => x.StageId).Select(s => new { StageId = s.Key, Data = s.Count() });
+            foreach (var item in data)
```
This might be replaced with correct select and cast functions

msvaillant commented on src/BaseOfTalents/DAL/Services/ReportService.cs:
```diff
@@ -0,0 +1,189 @@
+        {
+            // select all stages created  for the selected period
+            var stagesFilter = new List<Expression<Func<VacancyStageInfo, bool>>>();
+            stagesFilter.Add(x => x.CreatedOn > startDate && x.CreatedOn < endDate);
```
collection initializers, use it

msvaillant commented on src/BaseOfTalents/DAL/Services/ReportService.cs:
```diff
@@ -0,0 +1,189 @@
+
+            var result = new List<LocationsVacanciesReportDTO>();
+
+            foreach (var vacanciesGroup in vacanciesGroupedByLocations)
```
looping -> querying

msvaillant commented on src/BaseOfTalents/DAL/Services/ReportService.cs:
```diff
@@ -0,0 +1,189 @@
+
+            var result = new List<LocationDailyVacanciesReportDTO>();
+
+            foreach (var vacanciesGroup in vacanciesGroupedByLocations)
```
looping -> querying

msvaillant commented on src/BaseOfTalents/DAL/Services/ReportService.cs:
```diff
@@ -0,0 +1,189 @@
+            DateTime startDate,
+            DateTime? endDate)
+        {
+            var statesCreatedInCurrentPeriodFilter = new List<Expression<Func<VacancyState, bool>>>();
```
collection initializer

msvaillant commented on src/BaseOfTalents/DAL/Services/ReportService.cs:
```diff
@@ -0,0 +1,189 @@
+            ICollection<int> userIds,
+            DateTime? date)
+        {
+            var statesFilter = new List<Expression<Func<VacancyState, bool>>>();
```
collection initializer

msvaillant commented on src/BaseOfTalents/WebUI/Controllers/ReportsController.cs:
```diff
@@ -0,0 +1,81 @@
+        [Route("vacanciesReport")]
+        public IHttpActionResult GetDataForVacancyReport([FromBody]VacanciesReportParameters vacanciesReportParams)
+        {
+            if (ModelState.IsValid)
```
revert if statement

msvaillant commented on src/BaseOfTalents/WebUI/Controllers/ReportsController.cs:
```diff
@@ -0,0 +1,81 @@
+                        vacanciesReportParams.UserIds,
+                        vacanciesReportParams.StartDate
+                    );
+                if (vacanciesReportParams.EndDate != null)
```
minimize statement

msvaillant commented on src/BaseOfTalents/WebUI/app.config:
```diff
@@ -0,0 +1,3 @@
+<?xml version="1.0" encoding="utf-8"?>
```
this is not ok

msvaillant commented on src/BaseOfTalents/WebUI/WebUI.csproj:
```diff
@@ -15,7 +15,7 @@
     <AppDesignerFolder>Properties</AppDesignerFolder>
     <RootNamespace>WebUI</RootNamespace>
     <AssemblyName>WebUI</AssemblyName>
-    <TargetFrameworkVersion>v4.5.2</TargetFrameworkVersion>
```
not ok

msvaillant commented on src/BaseOfTalents/Domain/Domain.csproj:
```diff
@@ -9,8 +9,9 @@
     <AppDesignerFolder>Properties</AppDesignerFolder>
     <RootNamespace>Domain</RootNamespace>
     <AssemblyName>Domain</AssemblyName>
-    <TargetFrameworkVersion>v4.5.2</TargetFrameworkVersion>
```
not ok

msvaillant commented on src/BaseOfTalents/DAL/DAL.csproj:
```diff
@@ -9,10 +9,11 @@
     <AppDesignerFolder>Properties</AppDesignerFolder>
     <RootNamespace>DAL</RootNamespace>
     <AssemblyName>DAL</AssemblyName>
-    <TargetFrameworkVersion>v4.5.2</TargetFrameworkVersion>
```
not ok

msvaillant commented on src/BaseOfTalents/DAL/App.config:
```diff
@@ -1,17 +1,13 @@
   </entityFramework>
-</configuration>
\ No newline at end of file
+<startup><supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.5"/></startup></configuration>
```
what a hell?

horba commented on src/BaseOfTalents/DAL/DTO/ReportDTO/DailyVacanciesReportDTO.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace DAL.DTO.ReportDTO
+{
+    public class DailyVacanciesReportDTO : BaseEntityDTO
```
why inherite BaseEntityDTO?

horba commented on src/BaseOfTalents/DAL/DTO/ReportDTO/DailyVacanciesReportDTO.cs:
```diff
@@ -0,0 +1,17 @@
+    public class DailyVacanciesReportDTO : BaseEntityDTO
+    {
+        public int UserId { get; set; }
+        public string DisplayName { get; set; }
```
DisplayName of what? Report?

horba commented on src/BaseOfTalents/DAL/Services/ReportService.cs:
```diff
@@ -0,0 +1,189 @@
+            
+        private int GetVacanciesCountForUser(IEnumerable<VacancyState> states, int responsibleId, EntityState state)
+        {
+            return states.Where(x => x.State == state).Select(x => x.Vacancy).Where(x => x.ResponsibleId == responsibleId).Count();
```
.Where().Count() -> Count with lambda expression

kulish-alina committed:
ReportsController and ReportService were refactored

kulish-alina committed:
DisplayName was renamed to UserDisplayName in VacanciesReportDTO, UsersReportDTO, DailyVacanciesReportDTO,

ReportService

kulish-alina committed:
version of NETFramework was changed to v4.5.2
